### PR TITLE
Allow colon scripts to run with yarn foreach

### PR DESCRIPTION
**Pull request by crubier**
**What's the problem this PR addresses?**

Fix a bug in foreach.ts where the colon-named scripts are only evaluated in workspaces that explicitly list them

See https://github.com/yarnpkg/berry/issues/1933#issuecomment-728841786

**How did you fix it?**

Made a special case for when script contain a colon.

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

**Comment by arcanis** [comment]
Sounds good! Can you also add a test in `foreach.test.js`?

**Comment by crubier** [comment]
The problem I have is that the tests fail on master for some reason, so I can't really test it... Any idea why they fail locally ? 

```
> yarn test:integration foreach
 FAIL  packages/acceptance-tests/pkg-tests-specs/sources/commands/workspaces/foreach.test.js (16.794 s)
 Commands
    workspace foreach
      ✓ should run on current and descendant workspaces by default (1106 ms)
      ✓ should execute 'node' command (369 ms)
      ✓ should run scripts in parallel and interlace the output when run with --parallel --interlaced (2360 ms)
      ✓ should run scripts in parallel but following the topological order when run with --parallel --topological (2121 ms)
      ✓ should prefix the output with run with --verbose (1215 ms)
      ✓ should only run the scripts on workspaces that match the --include list (1010 ms)
      ✓ should never run the scripts on workspaces that match the --exclude list (1037 ms)
      ✓ should not fall into endless loop if foreach cmd is the same as lifecycle script name (1963 ms)
      ✓ should throw an error when using --jobs without --parallel (758 ms)
      ✓ should throw an error when using --jobs with a value lower than 2 (691 ms)
      ✓ can run on public workspaces only (955 ms)
      ✓ should return correct exit code when encountered errors in running scripts (980 ms)
      ✕ should include dependencies if using --recursive (705 ms)

  ● Commands › workspace foreach › should include dependencies if using --recursive

    expect(received).toMatchSnapshot()

    Snapshot name: `Commands workspace foreach should include dependencies if using --recursive 1`

    - Snapshot  - 5
    + Received  + 4

      Object {
    -   "code": 0,
    +   "code": 1,
        "stderr": "",
    -   "stdout": "➤ YN0000: Test Workspace A
    - ➤ YN0000: Test Workspace C
    - ➤ YN0000: Test Workspace B
    - ➤ YN0000: Done
    +   "stdout": "Unknown Syntax Error: Unsupported option name (\"--recursive\").
    +
    + $ yarn workspaces foreach [-A,--all] [-v,--verbose] [-p,--parallel] [-i,--interlaced] [-j,--jobs #0] [-t,--topological] [--topological-dev] [--include #0] [--exclude #0] [--private] <commandName> ...
      ",
      }

      436 |           }
      437 | 
    > 438 |           await expect({code, stdout, stderr}).toMatchSnapshot();
          |                                                ^
      439 |         }
      440 |       )
      441 |     );

      at makeTemporaryEnv (pkg-tests-specs/sources/commands/workspaces/foreach.test.js:438:48)

 › 1 snapshot failed.
Snapshot Summary
 › 1 snapshot failed from 1 test suite. Inspect your code changes or re-run jest with `-u` to update them.

Test Suites: 1 failed, 1 total
Tests:       1 failed, 12 passed, 13 total
Snapshots:   1 failed, 8 passed, 9 total
Time:        16.904 s
Ran all test suites matching /foreach/i.
```

**Comment by crubier** [comment]
Nevermind I was missing `yarn update-local`

**Comment by crubier** [comment]
Tests implemented, but I have weird results (see snapshots):

- The technique in https://github.com/yarnpkg/berry/issues/1933#issuecomment-728841786 does not work (it prints the same result for each package, not a different dir for each one)

**Comment by arcanis** [comment]
You made two small mistakes in your test:

- you put `test:foo` inside a workspace rather than the root; in itself it's not a problem, except when compounded with the second one
- instead of calling `yarn test:foo` (which itself would call `yarn workspaces foreach test:bar`), you called `yarn workspaces foreach test:foo`

As a result, your test was running `yarn test:foo` from each workspace, which were executed with a cwd set to `workspace-a`, so when `test:foo` tried to execute `yarn workspaces foreach test:bar` it only executed it on itself (since `foreach` is by default scoped to a workspace and its children, and `workspace-a` had none).

File: packages/acceptance-tests/pkg-tests-specs/sources/commands/workspaces/foreach.test.js [hunk 1]
```diff
@@ -1,3 +1,5 @@
+import {npath} from '@yarnpkg/fslib';
+
```
Comment (merceyz) [inline packages/acceptance-tests/pkg-tests-specs/sources/commands/workspaces/foreach.test.js:2]:
```suggestion
```

**Pull request merged**